router/router.go

```diff
@@ -181,6 +185,78 @@ func (self *Router) GetConnectEventsConfig() *env.ConnectEventsConfig {
 	return &self.config.ConnectEvents
 }
 
+// GetSANs retrieves the Subject Alternative Names (SANs) from a TLS certificate.
+func GetSANs(host string, port int) ([]string, error) {
+	address := fmt.Sprintf("%s:%d", host, port)
+	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: 5 * time.Second}, "tcp", address, &tls.Config{
+		InsecureSkipVerify: true, //only used to get certs, needs to be insecure
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect: %w", err)
+	}
+	defer conn.Close()
+
+	certs := conn.ConnectionState().PeerCertificates
+	if len(certs) == 0 {
+		return nil, fmt.Errorf("no certificates found")
+	}
+
+	sans := append([]string{}, certs[0].DNSNames...)
+	for _, ip := range certs[0].IPAddresses {
+		sans = append(sans, ip.String())
+	}
+
+	if len(sans) == 0 {
+		return nil, fmt.Errorf("no SANs found in certificate")
+	}
+
+	return sans, nil
+}
+
+func (self *Router) ProbeAdvertisements() {
+	for _, l := range self.config.Listeners {
+		if l.options["binding"] != "edge" {
+			continue
+		}
+
+		op := l.options["options"].(map[interface{}]interface{})
+		a := op["advertise"]
+		if a == nil {
+			continue
+		}
+
+		parts := strings.Split(a.(string), ":")
+		host := parts[0]
+		port, err := strconv.Atoi(parts[1])
+		if err != nil {
+			logrus.WithField("address", a).WithError(err).Debug("failed to parse advertise option")
+			continue
+		}
+
+		sans, serr := GetSANs(host, port)
+		if serr != nil {
+			logrus.WithField("host", host).WithError(serr).Warn("advertised address cannot be probed")
+			continue
+		}
+
+		found := false
+		for _, s := range sans {
+			if s == host {
+				found = true
+				break
+			}
+		}
```

Comment:
This doesn't correctly match all variants of SAN values.

DNS
- case insensitivity
- trailing dog ignore
- IDN Punycode conversions
IP
- ipv4 normalization
- ipv6 normalization

Reply:
If you have an `x509.Ceritificate`, it has a function called VerifyHostname(hostname string) that should handle all of this for you. I took a cursory look, it seems like a good fit. It uses the ip.Equal which handles IP normalization. DNS matching looks "fine" but has some caveats.

Reply:
used it. thx